kernel: send all fatal messages to the logging mechanism

tyrantlizard77 opened the pull request:
This PR:

- Removes the z_fatal_print() API
- All previous instances of z_fatal_print() replaced with LOG_ERR
- For test cases, a CONFIG_TEST_LOGGING_DEFAULTS has been introduced. It by default enables logging in immediate mode for all tests. Tests that need an alternate logging configuration (such as tests of the logger itself) may disable this in the project configuration.
- Fixes a few tests that fail if immediate logging is turned on

Since fatal errors only are reported to the logger, and fatal errors in CI logs should not have any dropped messages, we need to default to enabling the logger in immediate mode for all tests.

Longer term this paves the way to replace virtually all instances of printk() or customized debug macros in kernel or arch code with logging calls.

Commit:
kernel: remove z_fatal_print()

Use LOG_ERR instead.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: enable log in immediate mode by default

Fatal error reporting now only dumps to the log mechanism,
so enable it in immediate mode for all tests to ensure
that fatal errors are visible and no messages are lost.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
Some failures observed, looking into them:

bbc_microbit/tests/bluetooth/mesh/test_microbit - fails due to insufficient RAM space
qemu_x86_nommu/tests/kernel/profiling/profiling_api/kernel.profiling  - hangs during test
qemu_x86/tests/kernel/profiling/profiling_api/kernel.profiling - spurious interrupt exception
qemu_nios2/tests/crypto/tinycrypt/crypto.tinycrypt - fails due to insufficient ROM space

Commit:
tests: bluetooth: mesh: adjust microbit config

Disable logging defaults for this test, enabling logging
causes this test to not fit in RAM.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: profiling_api: increase idle stack size

This test for some reason wants to validate that
k_stack_analyze() works when called from the idle thread,
but with a default idle stack size of 256 this just results
in crashes.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

ceolin commented on arch/arc/core/fatal.c:
```diff
@@ -16,13 +16,14 @@
 #include <arch/cpu.h>
-#include <logging/log_ctrl.h>
+#include <logging/log.h>
+LOG_MODULE_DECLARE(os);
```
Don't this module need to be register at some point ?

andrewboie replied:
it is, in kernel/

ceolin replied:
indeed, somehow I didn't see it :/ it is registered in init.c

Commit:
board: qemu_nios2: set yaml RAM/ROM size

This emulated board has 128KB each of flash and SRAM.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

aescolar reviewed (dismissed):
A quite fundamental comment:
I'm quite concerned about the overall direction we are going: increasing dependencies in between subsystems and decreasing Zephyr's modularity. And, increasing requirements on the platform.
So even though in this case we are "just" routing the fatal prints thru the logger, and in itself it is not too bad, as one can just do without those prints, it seems to be a symptom of an overall tendency: An assumption that there is plenty of RAM and resources to spare for all possible fancy features. And an assumption that if you want to use the Zephyr kernel, you also want to use all that comes with Zephyr (and nothing else).

&
a relatively minor one (which I would still like fixed if this is to be merged):
In the 2nd commit message: "tests: enable log in immediate mode by default"
"Error reporting now only dumps to the log mechanism, so.." . I guess it is only "Fatal error reporting.."? If so, could you please add that "Fatal" word at the beginning of the line. Otherwise this commit message description could be quite misleading.

nordic-krch commented:
@aescolar I thought that we want to have single way of logging and in that case depending everywhere on single module rather than using alternatives is a good thing. Logger is used everywhere in zephyr then what sense does it make to not use it here where actually it is the most crucial when you think about feature like crashlog.

Current (printk) approach is maybe simpler but it's debugging only feature with console attached. Logger gives the same and adds more by supporting multiple backends. Once flash backend is added you can crashlog and that will be much more useful in the field than printk.

andrewboie commented:
> I'm quite concerned about the overall direction we are going: increasing dependencies in between subsystems and decreasing Zephyr's modularity. And, increasing requirements on the platform.
So even though in this case we are "just" routing the fatal prints thru the logger, and in itself it is not too bad, as one can just do without those prints, it seems to be a symptom of an overall tendency: An assumption that there is plenty of RAM and resources to spare for all possible fancy features. 

@aescolar right now we have a hodgepodge of whether code uses printk or logs to emit diagnostic data. Setting this up instead in a consistent way using a single interface is what we need to do. The logging subsystem is here, we should fully commit to it. And we already use log macros elsewhere in the kernel.

Worse case if this actually becomes a footprint issue for anyone: you could literally just macro LOG_ERR() to printk() and then there is no difference. log_strdup() would just be a no-op returning the supplied pointer. LOG_MODULE_DECLARE() expands to nothing. This interface can scale down to an implementation of trivial size if we need it to, controlled by Kconfig.

carlescufi commented:
I agree with the comments made by @andrewboie and @nordic-krch. Let's use a single mechanism, which will also allow for fatal errors to be stored in places like flash for devices in the field, and then optimize the logging subsystem if necessary for small footprint.

aescolar commented:
> Worst case if this actually becomes a footprint issue for anyone: you could literally just 

Correct, today, in this case.

> you could literally just macro LOG_ERR() to printk() 

(sidetracking a lot) Or you could use a tracing method which does not relay on having any debug strings in the embedded image, and logs or transfers out only the bare minimum amount of data to reconstruct the message on a PC.

Anyhow as I said that is an overall concern about the overall approach we are having in Zephyr, where this particular case is just one symptom.

@andrewboie About my second part of the review, do you agree about adding that "Fatal.." to the commit message?

@nordic-krch My concern is that the logger, even that it is a nice concept and quite powerful, how it is built today, it is way too resource consuming, and therefore you cannot assume it will be possible for those who build with Zephyr to be able to use it.
Zephyr is "just" the OS. The resources on the SOC are there for an application's purpose. Any extra KB the OS uses is one KB less for the application, to implement the features the user actually wants. The end user does not care about the logger. The logger is just a debugging facility for the developer, and therefore using it in a final image, in many cases would only be possible by trading off actual application functionality.

andrewboie commented:
> Correct, today, in this case.

I think you're waving your arms here. What logging interface additions or changes would prevent this?

>  Or you could use a tracing method which does not relay on having any debug strings in the embedded image, and logs or transfers out only the bare minimum amount of data to reconstruct the message on a PC.

You *CAN* with this. It's not implemented, but with this current interface we could:

1. Save the format strings in a special section that is stripped out of the binary and saved in a build file, we'll call it the index file.
2. At runtime, logging never dereferences the format string pointer or does the printk() logic, it just logs the pointer value of the format string and any arguments supplied.
3. After-the-fact postprocessing of the collected logs on a workstation will take the logged binary data, use the index file to fetch the format strings, and then turn them into fully formatted log strings.

We can also introduce additional instrumentation point types, what we have right now are events that are specified with printf like syntax, but we could also do counters, time intervals, even runtime histograms. I worked on a tracing system in grad school that did all this stuff and the concepts are adaptable to the infrastructure @nordic-krch has created.

> My concern is that the logger, even that it is a nice concept and quite powerful, how it is built today, it is way too resource consuming

You haven't established this, whereas I gave an example where we could have a stripped down implementation that is the same size as printk and a console driver. What's your counter-argument which shows that this isn't possible?

> About my second part of the review, do you agree about adding that "Fatal.." to the commit message?

I edited the commit message and pushed an update.

aescolar commented:
@andrewboie Thanks for correcting the commit message.

> You CAN with this..
@andrewboie I haven't said you can't, or have I? ;)
I was just presenting what I think would be a more optimal tracing/logging approach than what we have today, which you kindly elaborated on.

@andrewboie @nordic-krch 
Anyhow, with my comment I did not intend to say we should stop using the logger, but that we should not assume everybody can use it, as it is today.
And as I also said, this particular case is a minor thing, and, from my point of view it could go in.
My general comment in the review is about the general tendency, not about this case or the logger in particular.
I wouldn't call it arm waving, but raising a general concern, so others are made aware of it.

Anyhow, as you have corrected the commit message, I remove the -1, but I would think it would make sense to allow time for others to review it instead of hurrying it into 2.0. I presume there is no urgency here?

andrewboie commented:
> Anyhow, as you have corrected the commit message, I remove the -1, but I would think it would make sense to allow time for others to review it instead of hurrying it into 2.0. I presume there is no urgency here?

I don't have much of a need for this to land in 2.0. It would be nice though.

nashif commented:
@andrewboie can you rebase please?

nashif commented:
```
mac:zephyr(): gg -E "BT_ERR|BT_INF|BT_WRN|BT_DBG|LOG_" kernel/ arch/ | wc -l
      62
mac:zephyr(): gg -E "BT_ERR|BT_INF|BT_WRN|LOG_|BT_DBG" subsys/bluetooth/ | wc -l
    1909
```

nuff said :)

Commit:
tests: increase min_ram for two tests

These two tests were right at the knife edge of 16kb
on riscv64, and were not building with logging enabled
on that platform.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
huh, getting a weird link failure:

 Sanitycheck / qemu_x86_coverage:tests/subsys/jwt/libraries.encoding / qemu_x86_coverage/tests/subsys/jwt/libraries.encoding 

@pizi-nordic any idea? Think this is related to user mode log support

```
[178/187] Linking C executable zephyr/zephyr_prebuilt.elf
FAILED: : && ccache /opt/sdk/zephyr-sdk-0.10.3/i586-zephyr-elf/bin/i586-zephyr-elf-gcc    zephyr/CMakeFiles/zephyr_prebuilt.dir/misc/empty_file.c.obj  -o zephyr/zephyr_prebuilt.elf  -Wl,-T zephyr/linker.cmd -Wl,-Map=/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/out-3nd-pass/qemu_x86_coverage/tests/subsys/jwt/libraries.encoding/zephyr/zephyr_prebuilt.map -Wl,--whole-archive app/libapp.a zephyr/libzephyr.a zephyr/arch/arch/x86/core/libarch__x86__core.a zephyr/lib/libc/newlib/liblib__libc__newlib.a zephyr/soc/x86/ia32/libsoc__x86__ia32.a zephyr/subsys/testsuite/ztest/libsubsys__testsuite__ztest.a zephyr/subsys/jwt/libsubsys__jwt.a zephyr/drivers/serial/libdrivers__serial.a modules/mbedtls/lib..__modules__crypto__mbedtls.a -Wl,--no-whole-archive zephyr/kernel/libkernel.a zephyr/CMakeFiles/offsets.dir/arch/x86/core/offsets/offsets.c.obj -L"/opt/sdk/zephyr-sdk-0.10.3/i586-zephyr-elf/bin/../lib/gcc/i586-zephyr-elf/8.3.0" -L/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/out-3nd-pass/qemu_x86_coverage/tests/subsys/jwt/libraries.encoding/zephyr -lgcc -Wl,--print-memory-usage -Wl,--gc-sections -Wl,--build-id=none -Wl,--sort-common=descending -Wl,--sort-section=alignment -u_OffsetAbsSyms -u_ConfigAbsSyms -nostdlib -static -no-pie -Wl,-X -Wl,-N -Wl,--orphan-handling=warn -Wl,--fatal-warnings -lm -lc -lgcc -L"/opt/sdk/zephyr-sdk-0.10.3/i586-zephyr-elf/i586-zephyr-elf"/lib/. && :
Memory region         Used Size  Region Size  %age Used
             ROM:      266276 B         4 MB      6.35%
             RAM:      299244 B         4 MB      7.13%
        IDT_LIST:         416 B         2 KB     20.31/opt/sdk/zephyr-sdk-0.10.3/i586-zephyr-elf/bin/../lib/gcc/i586-zephyr-elf/8.3.0/../../../../i586-zephyr-elf/bin/ld: zephyr/libzephyr.a(log_core.c.obj): in function `log_generic':
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/subsys/logging/log_core.c:390: undefined reference to `log_generic_from_user'
/opt/sdk/zephyr-sdk-0.10.3/i586-zephyr-elf/bin/../lib/gcc/i586-zephyr-elf/8.3.0/../../../../i586-zephyr-elf/bin/ld: zephyr/arch/arch/x86/core/libarch__x86__core.a(fatal.c.obj): in function `dump_entry_flags':
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/arch/x86/core/ia32/fatal.c:278: undefined reference to `log_from_user'
/opt/sdk/zephyr-sdk-0.10.3/i586-zephyr-elf/bin/../lib/gcc/i586-zephyr-elf/8.3.0/../../../../i586-zephyr-elf/bin/ld: zephyr/arch/arch/x86/core/libarch__x86__core.a(fatal.c.obj): in function `unwind_stack':
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/arch/x86/core/ia32/fatal.c:81: undefined reference to `log_from_user'
/opt/sdk/zephyr-sdk-0.10.3/i586-zephyr-elf/bin/../lib/gcc/i586-zephyr-elf/8.3.0/../../../../i586-zephyr-elf/bin/ld: /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/arch/x86/core/ia32/fatal.c:87: undefined reference to `log_from_user'
/opt/sdk/zephyr-sdk-0.10.3/i586-zephyr-elf/bin/../lib/gcc/i586-zephyr-elf/8.3.0/../../../../i586-zephyr-elf/bin/ld: /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/arch/x86/core/ia32/fatal.c:109: undefined reference to `log_from_user'
/opt/sdk/zephyr-sdk-0.10.3/i586-zephyr-elf/bin/../lib/gcc/i586-zephyr-elf/8.3.0/../../../../i586-zephyr-elf/bin/ld: zephyr/arch/arch/x86/core/libarch__x86__core.a(fatal.c.obj): in function `dump_page_fault':
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/arch/x86/core/ia32/fatal.c:309: undefined reference to `log_from_user'
/opt/sdk/zephyr-sdk-0.10.3/i586-zephyr-elf/bin/../lib/gcc/i586-zephyr-elf/8.3.0/../../../../i586-zephyr-elf/bin/ld: zephyr/arch/arch/x86/core/libarch__x86__core.a(fatal.c.obj):/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/arch/x86/core/ia32/fatal.c:311: more undefined references to `log_from_user' follow
collect2: error: ld returned 1 exit status
%
ninja: build stopped: subcommand failed.
```

pizi-nordic commented:
I seen such errors only when both userspace and optimization in compiler were disabled. The object files held reference for never-called symbols. If you set CONFIG_NO_OPTIMIZATION, the stubs will be added for the missing functions.

andrewboie commented:
> I seen such errors only when both userspace and optimization in compiler were disabled. The object files held reference for never-called symbols. If you set CONFIG_NO_OPTIMIZATION, the stubs will be added for the missing functions.

@pizi-nordic Right but we have been gathering coverage data with user mode and system calls for quite some time now without issues.
This seems a problem specific to the user mode logging hooks.
Since before this patch, very few tests turned logging on, I think this is a problem with the user mode logging patch.

pizi-nordic commented:
So what about the following change:
```
diff --git a/subsys/logging/log_core.c b/subsys/logging/log_core.c                                      
index 3828c98241..80bfd6c52b 100644                                                                     
--- a/subsys/logging/log_core.c                                                                         
+++ b/subsys/logging/log_core.c                                                                         
@@ -1081,7 +1081,7 @@ void log_hexdump_from_user(struct log_msg_ids src_level, const char *metadata,    
        src_level_union.structure = src_level;                                                          
        z_log_hexdump_from_user(src_level_union.value, metadata, data, len);                            
 }                                                                                                      
-#elif defined(CONFIG_NO_OPTIMIZATIONS)                                                                 
+#else                                                                                                  
 /*                                                                                                     
  * With the optimizations disabled, the dead code is not removed and                                   
  * references to functions used only by the usermode remains in the                                    
```

This will provide stubs for usermode similar to stubs for normal logging which are there even if logger is disabled.

andrewboie commented:
> So what about the following change:

I'll try this...I need to understand better why the stubs are needed (this went in when I was on vacation and I'm just now looking at the code) but I'll see if that works.

andrewboie commented:
Yeah, I see it now, that's the right fix. I'll add to the series.

Commit:
log_core: always build stubs when !userspace

These functions get called from various places and
we were observing linker errors. Always build the stubs
when userspace is disabled, gc-sections will discard them
if unused anyway.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>
